Simplify the implementation of "dedup" semi-join paths.

hlinnaka opened the pull request:
There are two parts to this commit:

1. Instead of using ctids and "fake" ctids to disambiguate rows that have
   been duplicated by a Broadcast motion, generate a synthetic 64-bit
   rowid specifically for that purpose, just below the Broadcast. The old
   method had to generate subplan IDs, fake ctids etc. to have a
   combination of columns that formed the unique ID, and that entailed a
   lot of infrastructure, like the whole concept of "pseudo columns". The
   new method is much simpler, and it works the same way regardless of
   what's below the Broadcast. The new 64-bit rowids are generated by
   a new RowIdExpr expression node. The high 16 bits of the rowid are
   the segment ID, and the low bits are generated by a simple counter.

2. Replace the cdbpath_dedup_fixup() post-processing step on the Plan
   tree, by adding the unique ID column (now a RowIdExpr) in the Path's
   pathtarget in the first place. cdbpath_motion_for_join() is the
   function that does that now. Seems like a logical place, since that's
   where any needed Motions are added on top of the join inputs. The
   responsibility for creating the Unique path is moved from
   cdb_add_join_path() to the create_nestloop/hashjoin/mergejoin_path()
   functions.

NOTE: This PR contains two commits. The first one is the same as in https://github.com/greenplum-db/gpdb/pull/9702. Please ignore that one here, and leave comments on that on the other PR. The focus of this PR is the second commit. (I was hacking on this refactoring when I realized the pre-existing bug that the first commit fixes.)

hlinnaka commented:
Pushed the prerequisite commit, so just one commit to review here now.

kainwen commented:
Hi, I am going to review this. Just scan the code and get the outline. Will spend more time on it.

Will finish reviewing soon.

kainwen commented on src/test/regress/expected/join.out:
```diff
@@ -5494,13 +5494,13 @@ select * from
   (select b.q1 as bq1, c.q1 as cq1, least(a.q1,b.q1,c.q1) from
    int8_tbl b cross join int8_tbl c) ss
   on a.q2 = ss.bq1;
-ERROR:  could not devise a query plan for the given query (pathnode.c:467)
```
This diff is not very interesting.

Does this mean every time we modify the code of `pathnode.c`, we need update this file?

I am not sure if we can use Regexp to skip this kind of check.

hlinnaka replied:
> Does this mean every time we modify the code of `pathnode.c`, we need update this file?

Unfortunately, yes.

> I am not sure if we can use Regexp to skip this kind of check.

Yeah. There's a FIXME comment above this, this was committed this way as a temporary fix. But no one's gotten around to take a proper look at the case and fix it.

kainwen replied:
> Yeah. There's a FIXME comment above this, this was committed this way as a temporary fix. But no one's gotten around to take a proper look at the case and fix it.

OK. Let's handle this in future.

kainwen commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -3418,7 +3362,16 @@ create_nestloop_path(PlannerInfo *root,
 	final_cost_nestloop(root, pathnode, workspace, sjinfo, semifactors);
 
-	return pathnode;
+	if (counter_id)
```
I am OK with this.

I am not sure if we check jointype here is better. And in the if branch, add assert(counter_id>0).

kainwen replied:
Also, this piece of code is copied in three places. They are the same.

Would it be better to wrap the code in a helper function? Just my two cents.

hlinnaka replied:
> I am not sure if we check jointype here is better. And in the if branch, add assert(counter_id>0).

Ok, changed that way. I put the assertion inside create_unique_rowid_path().

> Also, this piece of code is copied in three places. They are the same.
> Would it be better to wrap the code in a helper function? Just my two cents.

I think it's better this way. Yeah, it's a bit repetitive, but it's pretty much just a function call. The helper function would be a function call too, so I don't think it would really help much.

In fact, we used to add this Unique path in cdb_add_join_path() before this PR. I tried to keep it that way at first, but it had its own problems. We'd need to somehow pass the RowIdExpr ID from here up to the caller, and to cdb_add_join_path(). Maybe store it in the JoinPath struct. But overall, I think this is more clear.

kainwen commented on src/backend/executor/execQual.c:
```diff
@@ -6220,6 +6247,19 @@ ExecInitExpr(Expr *node, PlanState *parent)
+				RowIdExprState *exprstate = makeNode(RowIdExprState);
+
+				/* FIXME: need something else to make unique? */
+				exprstate->counter = ((uint64) GpIdentity.dbid) << 48 ;
```
So the max rows for the subplan is 2^48 which is a huge number.

Because the planner choose this plan which means the rel should not contain so many rows(because we broadcast them).

Do I understand correctly?

hlinnaka replied:
Yeah. It seems far-fetched in general to process that many rows in a single query.

kainwen commented:
I do not know why the pr pipeline is not triggered. We should run full ICW before pushing this.

hlinnaka commented:
Rebased and changed the if() clauses in pathnode.c per your suggestion. I also renamed "counter_id" fields to "rowidexpr_id", that seems more clear. And I also renamed "counter" in RowIdExprState to "rowcounter".

Let's see if the PR pipeline kicks in now.

kainwen commented:
> Rebased and changed the if() clauses in pathnode.c per your suggestion. I also renamed "counter_id" fields to "rowidexpr_id", that seems more clear. And I also renamed "counter" in RowIdExprState to "rowcounter".
> 
> Let's see if the PR pipeline kicks in now.

Great. Pipeline is triggered. I think we can push this after pr pipeline is green.

Last question:

Do we need to remove function scan's fake ctid introduced in the commit https://github.com/greenplum-db/gpdb/commit/d8886cf9cf09b4148ebc9812e6b0663655c47f77
 to fix the issue https://github.com/greenplum-db/gpdb/issues/9159.

Thanks.

kainwen commented:
> Rebased and changed the if() clauses in pathnode.c per your suggestion. I also renamed "counter_id" fields to "rowidexpr_id", that seems more clear. And I also renamed "counter" in RowIdExprState to "rowcounter".
> 
> Let's see if the PR pipeline kicks in now.

Sorry I search the code do not find the logic of function scan ctid. It seems removed.

Ignore my previous noise.

hlinnaka commented:
On 11/03/2020 17:22, Zhenghua Lyu wrote:
> Do we need to remove function scan's fake ctid introduced in the commit 
> d8886cf 
> <https://github.com/greenplum-db/gpdb/commit/d8886cf9cf09b4148ebc9812e6b0663655c47f77>
> to fix the issue #9159 <https://github.com/greenplum-db/gpdb/issues/9159>.
Yeah, that is being removed in this PR.

But now that I look closer, the test case that was added in that commit 
no longer produces the DEDUP plan as it before this PR. Before this PR, 
the plan was:

postgres=# explain select count(*) from pg_backend_pid() b(a) where b.a 
% 100000 in (select a from t1);
                                                     QUERY PLAN 

-------------------------------------------------------------------------------------------------------------------
  Finalize Aggregate  (cost=1361.74..1361.75 rows=1 width=8)
    ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=1361.70..1361.73 
rows=1 width=8)
          ->  Partial Aggregate  (cost=1361.70..1361.71 rows=1 width=8)
                ->  Result  (cost=1361.27..1361.66 rows=6 width=0)
                      ->  HashAggregate  (cost=1361.27..1361.66 rows=6 
width=6)
                            Group Key: b.ctid
                            ->  Redistribute Motion 3:3  (slice2; 
segments: 3)  (cost=0.02..1361.52 rows=6 width=6)
                                  Hash Key: b.ctid
                                  ->  Hash Join  (cost=0.02..1361.19 
rows=6 width=6)
                                        Hash Cond: (t1.a = (b.a % 100000))
                                        ->  Seq Scan on t1 
(cost=0.00..1111.00 rows=33334 width=4)
                                        ->  Hash  (cost=0.01..0.01 
rows=1 width=10)
                                              ->  Function Scan on b 
(cost=0.00..0.01 rows=1 width=10)
  Optimizer: Postgres query optimizer
(14 rows)

But now you get an upstream-style JOIN_OUTER_UNIQUE plan instead:

                                        QUERY PLAN 

----------------------------------------------------------------------------------------
  Finalize Aggregate  (cost=1944.60..1944.61 rows=1 width=8)
    ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=1944.57..1944.60 
rows=1 width=8)
          ->  Partial Aggregate  (cost=1944.57..1944.58 rows=1 width=8)
                ->  Hash Join  (cost=1361.02..1944.52 rows=6 width=0)
                      Hash Cond: (t1.a = (b.a % 100000))
                      ->  HashAggregate  (cost=1361.00..1694.33 
rows=33334 width=4)
                            Group Key: t1.a
                            ->  Seq Scan on t1  (cost=0.00..1111.00 
rows=33334 width=4)
                      ->  Hash  (cost=0.01..0.01 rows=1 width=4)
                            ->  Function Scan on b  (cost=0.00..0.01 
rows=1 width=4)
  Optimizer: Postgres query optimizer
(11 rows)

This is an interesting case. The new plan is estimated to be more 
expensive than the old one. The reason we no longer get the old plan is 
because the planner doesn't consider that plan anymore. The reason is 
that the Function Scan has General locus, and my PR only consider DEDUP 
plans when both sides of the join are partitioned.

There's a comment above create_unique_rowid_path() (that I wrote some 
time ago according to git history) that says:

  * The JOIN_DEDUP_SEMI plan will look something like this:
  *
  * postgres=# explain select * from s where exists (select 1 from r 
where s.a = r.b);
  *                                                   QUERY PLAN 

  * 
---------------------------------------------------------------------------------------------------------------
  *  Gather Motion 3:1  (slice1; segments: 3)  (cost=153.50..155.83 
rows=100 width=8)
  *    ->  HashAggregate  (cost=153.50..153.83 rows=34 width=8)
  *          Group Key: (RowIdExpr)
  *          ->  Redistribute Motion 3:3  (slice2; segments: 3) 
(cost=11.75..153.00 rows=34 width=8)
  *                Hash Key: (RowIdExpr)
  *                ->  Hash Join  (cost=11.75..151.00 rows=34 width=8)
  *                      Hash Cond: (r.b = s.a)
  *                      ->  Seq Scan on r  (cost=0.00..112.00 rows=3334 
width=4)
  *                      ->  Hash  (cost=8.00..8.00 rows=100 width=8)
  *                            ->  Broadcast Motion 3:3  (slice3; 
segments: 3)  (cost=0.00..8.00 rows=100 width=8)
  *                                  ->  Seq Scan on s  (cost=0.00..4.00 
rows=34 width=8)
  *  Optimizer: Postgres query optimizer
  * (12 rows)
  *
  * In PostgreSQL, this is never better than doing a JOIN_SEMI directly.
  * But it can be a win in GPDB, if the distribution of the outer and inner
  * relations don't match, and the outer relation is much larger than the
  * inner relation.

I took that "In PostgreSQL, this is never better than doing a JOIN_SEMI 
directly" comment as gospel in this PR, and disabled this join type 
altogether when one side of the join is replicated or general so that 
the join doesn't require redistributing the inputs. But this test query 
is an example of a case where the JOIN_DEDUP plan is still cheaper! 
Clearly the comment needs to be adjusted.

To enable that plan again, at first I tried to just remove this 
locus-check here:

	case JOIN_DEDUP_SEMI:
		/*
		 * Skip rowid unique path if distinct rels are replicated tables.
		 * A normal semi-join plan is always better in that case.
		 */
		if (!CdbPathLocus_IsPartitioned(inner.locus))
			goto fail;
		if (!CdbPathLocus_IsPartitioned(outer.locus))
			goto fail;
		inner.ok_to_replicate = false;
		outer.path = add_rowid_to_path(root, outer.path, p_rowidexpr_id);
		*p_outer_path = outer.path;
		break;

But that's not quite right, and resulted in test failures in queries 
involving replicated tables. The corresponding check before this PR was:

-               /*
-                * Skip rowid unique path if distinct rels are 
replicated tables
-                * The reason is ctid + gp_segment_id can not identify a 
logical
-                * row of replicated table.
-                *
-                * TODO: add a motion on top of segmentGeneral node to 
support
-                * rowid unique.
-                */
-               if (CdbPathLocus_IsPartitioned(path->locus) &&
-                   CdbPathLocus_IsSegmentGeneral(((JoinPath 
*)path)->outerjoinpath->locus))
-                       return;

But that was still not quite right. The old coding had a very subtle 
assumption regarding Function Scans with General locus: we assumed that 
when multiple segments execute a Function Scan (which naturally mustn't 
be volatile), the function returns the rows in the same order in all 
segments, and therefore each segment assigns the same "synthetic ctid" 
on each logically equivalent row. To have the same effect with the 
RowIdExpr node in this PR, I guess we need to have two kinds of 
RowIdExprs. Normally, each segment needs to produce unique row-IDs, and 
we include the segment ID in the row counter for that reason. But when 
the RowIdExpr is used on a General path, the node should produce the 
same sequence on each segment, so the segment ID should not be included 
in it. And I believe we still need to forbid SegmentGeneral paths; a 
scan on a replicated table might produce rows in different order.

Let me re-iterate the assumptions made previously: A path with General 
locus produces the same rows, *in same order*, on all segments. A path 
with SegmentGeneral locus also produces same rows on all segments, but 
not necessarily in same order.

I think we can put back that logic, but I also wonder if that really was 
a safe assumption. And even if it was, whether it's worth keeping it; 
this seems like a rare case in practice.

- Heikki

kainwen commented:
> On 11/03/2020 17:22, Zhenghua Lyu wrote: Do we need to remove function scan's fake ctid introduced in the commit [d8886cf](https://github.com/greenplum-db/gpdb/commit/d8886cf9cf09b4148ebc9812e6b0663655c47f77) <[d8886cf](https://github.com/greenplum-db/gpdb/commit/d8886cf9cf09b4148ebc9812e6b0663655c47f77)> to fix the issue #9159 <#9159>.
> Yeah, that is being removed in this PR. But now that I look closer, the test case that was added in that commit no longer produces the DEDUP plan as it before this PR. Before this PR, the plan was: postgres=# explain select count(*) from pg_backend_pid() b(a) where b.a % 100000 in (select a from t1); QUERY PLAN
> […](#)
> ------------------------------------------------------------------------------------------------------------------- Finalize Aggregate (cost=1361.74..1361.75 rows=1 width=8) -> Gather Motion 3:1 (slice1; segments: 3) (cost=1361.70..1361.73 rows=1 width=8) -> Partial Aggregate (cost=1361.70..1361.71 rows=1 width=8) -> Result (cost=1361.27..1361.66 rows=6 width=0) -> HashAggregate (cost=1361.27..1361.66 rows=6 width=6) Group Key: b.ctid -> Redistribute Motion 3:3 (slice2; segments: 3) (cost=0.02..1361.52 rows=6 width=6) Hash Key: b.ctid -> Hash Join (cost=0.02..1361.19 rows=6 width=6) Hash Cond: (t1.a = (b.a % 100000)) -> Seq Scan on t1 (cost=0.00..1111.00 rows=33334 width=4) -> Hash (cost=0.01..0.01 rows=1 width=10) -> Function Scan on b (cost=0.00..0.01 rows=1 width=10) Optimizer: Postgres query optimizer (14 rows) But now you get an upstream-style JOIN_OUTER_UNIQUE plan instead: QUERY PLAN
> ---------------------------------------------------------------------------------------- Finalize Aggregate (cost=1944.60..1944.61 rows=1 width=8) -> Gather Motion 3:1 (slice1; segments: 3) (cost=1944.57..1944.60 rows=1 width=8) -> Partial Aggregate (cost=1944.57..1944.58 rows=1 width=8) -> Hash Join (cost=1361.02..1944.52 rows=6 width=0) Hash Cond: (t1.a = (b.a % 100000)) -> HashAggregate (cost=1361.00..1694.33 rows=33334 width=4) Group Key: t1.a -> Seq Scan on t1 (cost=0.00..1111.00 rows=33334 width=4) -> Hash (cost=0.01..0.01 rows=1 width=4) -> Function Scan on b (cost=0.00..0.01 rows=1 width=4) Optimizer: Postgres query optimizer (11 rows) This is an interesting case. The new plan is estimated to be more expensive than the old one. The reason we no longer get the old plan is because the planner doesn't consider that plan anymore. The reason is that the Function Scan has General locus, and my PR only consider DEDUP plans when both sides of the join are partitioned. There's a comment above create_unique_rowid_path() (that I wrote some time ago according to git history) that says: * The JOIN_DEDUP_SEMI plan will look something like this: * * postgres=# explain select * from s where exists (select 1 from r where s.a = r.b); * QUERY PLAN * --------------------------------------------------------------------------------------------------------------- * Gather Motion 3:1 (slice1; segments: 3) (cost=153.50..155.83 rows=100 width=8) * -> HashAggregate (cost=153.50..153.83 rows=34 width=8) * Group Key: (RowIdExpr) * -> Redistribute Motion 3:3 (slice2; segments: 3) (cost=11.75..153.00 rows=34 width=8) * Hash Key: (RowIdExpr) * -> Hash Join (cost=11.75..151.00 rows=34 width=8) * Hash Cond: (r.b = s.a) * -> Seq Scan on r (cost=0.00..112.00 rows=3334 width=4) * -> Hash (cost=8.00..8.00 rows=100 width=8) * -> Broadcast Motion 3:3 (slice3; segments: 3) (cost=0.00..8.00 rows=100 width=8) * -> Seq Scan on s (cost=0.00..4.00 rows=34 width=8) * Optimizer: Postgres query optimizer * (12 rows) * * In PostgreSQL, this is never better than doing a JOIN_SEMI directly. * But it can be a win in GPDB, if the distribution of the outer and inner * relations don't match, and the outer relation is much larger than the * inner relation. I took that "In PostgreSQL, this is never better than doing a JOIN_SEMI directly" comment as gospel in this PR, and disabled this join type altogether when one side of the join is replicated or general so that the join doesn't require redistributing the inputs. But this test query is an example of a case where the JOIN_DEDUP plan is still cheaper! Clearly the comment needs to be adjusted. To enable that plan again, at first I tried to just remove this locus-check here: case JOIN_DEDUP_SEMI: /* * Skip rowid unique path if distinct rels are replicated tables. * A normal semi-join plan is always better in that case. */ if (!CdbPathLocus_IsPartitioned(inner.locus)) goto fail; if (!CdbPathLocus_IsPartitioned(outer.locus)) goto fail; inner.ok_to_replicate = false; outer.path = add_rowid_to_path(root, outer.path, p_rowidexpr_id); *p_outer_path = outer.path; break; But that's not quite right, and resulted in test failures in queries involving replicated tables. The corresponding check before this PR was: - /* - * Skip rowid unique path if distinct rels are replicated tables - * The reason is ctid + gp_segment_id can not identify a logical - * row of replicated table. - * - * TODO: add a motion on top of segmentGeneral node to support - * rowid unique. - */ - if (CdbPathLocus_IsPartitioned(path->locus) && - CdbPathLocus_IsSegmentGeneral(((JoinPath *)path)->outerjoinpath->locus)) - return; But that was still not quite right. The old coding had a very subtle assumption regarding Function Scans with General locus: we assumed that when multiple segments execute a Function Scan (which naturally mustn't be volatile), the function returns the rows in the same order in all segments, and therefore each segment assigns the same "synthetic ctid" on each logically equivalent row. To have the same effect with the RowIdExpr node in this PR, I guess we need to have two kinds of RowIdExprs. Normally, each segment needs to produce unique row-IDs, and we include the segment ID in the row counter for that reason. But when the RowIdExpr is used on a General path, the node should produce the same sequence on each segment, so the segment ID should not be included in it. And I believe we still need to forbid SegmentGeneral paths; a scan on a replicated table might produce rows in different order. Let me re-iterate the assumptions made previously: A path with General locus produces the same rows, *in same order*, on all segments. A path with SegmentGeneral locus also produces same rows on all segments, but not necessarily in same order. I think we can put back that logic, but I also wonder if that really was a safe assumption. And even if it was, whether it's worth keeping it; this seems like a rare case in practice. - Heikki

Thanks for your detailed explanation.  Those cases should be rare in practical env.

I prefer not optimizing those case.  But I am still OK with putting back the logic.

hlinnaka commented:
> Thanks for your detailed explanation. Those cases should be rare in practical env.
> 
> I prefer not optimizing those case. But I am still OK with putting back the logic.

I realized that there's another way we could support General, and also SegmentGeneral paths. Even though the input is available on all segments, we can force it to run one a single segment. Then we can reliably generate the row IDs. After generating the row IDs, it needs to be broadcast/redistributed, but I think that's OK. The nice thing about that is that it also works with replicated tables (SegmentGeneral). The possibility of doing that was in fact already mentioned in a TODO comment, which this PR removed.

I added a commit to this PR to implement that, and to also add a bunch of tests. Please take look at that second commit. I'll squash it before pushing, but I left it as a separate commit for easier review until then.

While hacking on this, I bumped into https://github.com/greenplum-db/gpdb/issues/9741. This coincidentally fixes that too (the bug was that the ctid was not always generated in Function Scans, but with this PR we don't rely on the synthetic ctids anymore).

kainwen commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1269,33 +1269,75 @@ cdbpath_motion_for_join(PlannerInfo *root,
+					 CdbPathLocus_IsSegmentGeneral(outer.locus))
+			{
+				CdbPathLocus_MakeSingleQE(&outer.locus,
+										  CdbPathLocus_NumSegments(inner.locus));
```
For segment general, this should be `Min(inner.locus, outer.locus)`.

kainwen commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1269,33 +1269,75 @@ cdbpath_motion_for_join(PlannerInfo *root,
+					 CdbPathLocus_IsSegmentGeneral(inner.locus))
+			{
+				CdbPathLocus_MakeSingleQE(&inner.locus,
+										  CdbPathLocus_NumSegments(outer.locus));
```
Same as the above comments.

kainwen commented:
> > Thanks for your detailed explanation. Those cases should be rare in practical env.
> > I prefer not optimizing those case. But I am still OK with putting back the logic.
> 
> I realized that there's another way we could support General, and also SegmentGeneral paths. Even though the input is available on all segments, we can force it to run one a single segment. Then we can reliably generate the row IDs. After generating the row IDs, it needs to be broadcast/redistributed, but I think that's OK. The nice thing about that is that it also works with replicated tables (SegmentGeneral). The possibility of doing that was in fact already mentioned in a TODO comment, which this PR removed.
> 
> I added a commit to this PR to implement that, and to also add a bunch of tests. Please take look at that second commit. I'll squash it before pushing, but I left it as a separate commit for easier review until then.
> 
> While hacking on this, I bumped into #9741. This coincidentally fixes that too (the bug was that the ctid was not always generated in Function Scans, but with this PR we don't rely on the synthetic ctids anymore).

The method looks good to me.

kainwen commented:
Both orca and planner icw failed in the PR pipeline for the same test case:

```
======================================================================
DIFF FILE: ../gpdb_src/src/test/regress/regression.diffs
----------------------------------------------------------------------

--- /tmp/build/b235ddbb/gpdb_src/src/test/regress/expected/pg_stat.out	2020-03-13 23:05:40.053651792 +0000
+++ /tmp/build/b235ddbb/gpdb_src/src/test/regress/results/pg_stat.out	2020-03-13 23:05:40.061651790 +0000
@@ -63,7 +63,7 @@
 from pg_stat_all_tables where relname = 'pg_stat_test';
  schemaname |   relname    | seq_scan | seq_tup_read | idx_scan | idx_tup_fetch | n_tup_ins | n_tup_upd | n_tup_del | n_tup_hot_upd | n_live_tup | n_dead_tup 
 ------------+--------------+----------+--------------+----------+---------------+-----------+-----------+-----------+---------------+------------+------------
- public     | pg_stat_test |        4 |          391 |        1 |             0 |       110 |         0 |        19 |             0 |         91 |         19
+ public     | pg_stat_test |        4 |          391 |        0 |             0 |       110 |         0 |        16 |             0 |         94 |         16
 (1 row)
 
 select
@@ -72,7 +72,7 @@
 from pg_stat_user_tables where relname = 'pg_stat_test';
  schemaname |   relname    | seq_scan | seq_tup_read | idx_scan | idx_tup_fetch | n_tup_ins | n_tup_upd | n_tup_del | n_tup_hot_upd | n_live_tup | n_dead_tup 
 ------------+--------------+----------+--------------+----------+---------------+-----------+-----------+-----------+---------------+------------+------------
- public     | pg_stat_test |        4 |          391 |        1 |             0 |       110 |         0 |        19 |             0 |         91 |         19
+ public     | pg_stat_test |        4 |          391 |        0 |             0 |       110 |         0 |        16 |             0 |         94 |         16
 (1 row)
 
 select
@@ -80,7 +80,7 @@
 from pg_stat_all_indexes where relname = 'pg_stat_test';
  schemaname |   relname    |       indexrelname       | idx_scan | idx_tup_read | idx_tup_fetch 
 ------------+--------------+--------------------------+----------+--------------+---------------
- public     | pg_stat_test | pg_stat_user_table_index |        1 |            1 |             0
+ public     | pg_stat_test | pg_stat_user_table_index |        0 |            0 |             0
 (1 row)
 
 select
@@ -88,7 +88,7 @@
 from pg_stat_user_indexes where relname = 'pg_stat_test';
  schemaname |   relname    |       indexrelname       | idx_scan | idx_tup_read | idx_tup_fetch 
 ------------+--------------+--------------------------+----------+--------------+---------------
- public     | pg_stat_test | pg_stat_user_table_index |        1 |            1 |             0
+ public     | pg_stat_test | pg_stat_user_table_index |        0 |            0 |             0
 (1 row)
 
 reset optimizer;
```

This is a known issue. That the `pg_stat` info is updated not in time. We should fix this flaky case in another PR. I have recorded it in a story.

hlinnaka commented:
Pushed, thanks for the review!
